src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -300,11 +279,94 @@ private ValueTask<Optional<TElement>> TryDequeueCoreAsync()
                         _waiters = new Queue<TaskCompletionSource<Optional<TElement>>>();
                     }
 
-                    var waiter = new TaskCompletionSource<Optional<TElement>>();
+                    // Continuations run asynchronously to ensure user code does not execute within protected regions.
+                    var waiter = new TaskCompletionSource<Optional<TElement>>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    AttachCancellation(waiter, cancellationToken);
                     _waiters.Enqueue(waiter);
                     return new ValueTask<Optional<TElement>>(waiter.Task);
                 }
             }
         }
+
+        /// <summary>
+        /// Cancels a <see cref="TaskCompletionSource{TResult}.Task"/> if a given <see cref="CancellationToken"/> is canceled.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by a successfully completed <see cref="Task{TResult}"/>.</typeparam>
+        /// <param name="taskCompletionSource">The <see cref="TaskCompletionSource{TResult}"/> to cancel.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
+        /// <seealso href="https://github.com/microsoft/vs-threading/blob/558f24c576cc620a00b20ed1fa90a5e2d13b0440/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs#L181-L255"/>
+        private static void AttachCancellation<T>(TaskCompletionSource<T> taskCompletionSource, CancellationToken cancellationToken)
+        {
+            _ = taskCompletionSource ?? throw new ArgumentNullException(nameof(taskCompletionSource));
+
+            if (!cancellationToken.CanBeCanceled || taskCompletionSource.Task.IsCompleted)
+                return;
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                taskCompletionSource.TrySetCanceled(cancellationToken);
+                return;
+            }
+
+            var tuple = new CancelableTaskCompletionSource<T>(taskCompletionSource, cancellationToken);
```

Comment:
Why the name `tuple`? This isn't a tuple.

Reply:
No idea. The same appears in the original code 😆

Reply:
➡️ Renamed the local